packages/flutter/lib/src/foundation/timeline.dart

```diff
@@ -0,0 +1,395 @@
+    startSync(name, arguments: arguments, flow: flow);
+    try {
+      return function();
+    } finally {
+      finishSync();
+    }
+  }
+
+  /// The current time stamp from the clock used by the timeline in
+  /// microseconds.
+  ///
+  /// When run on the Dart VM, uses the same monotonic clock as the embedding
+  /// API's `Dart_TimelineGetMicros`.
+  ///
+  /// When run on the web, uses `window.performance.now`.
+  static int get now => impl.performanceTimestamp.toInt();
+
+  /// Returns timings collected since [collectionEnabled] was set to true, or
+  /// since the previous [reset], whichever was last.
+  static AggregatedTimings collect() {
+    if (!_collectionEnabled) {
+      throw StateError('Timeline metric collection not enabled.');
+    }
+    return AggregatedTimings(_buffer.timings);
+  }
+
+  /// Forgets all previously collected timing data.
+  ///
+  /// This method can be used to break up the data by frames. To do that, call
+  /// [collect] at the end of the frame, then call [reset] so that the next
+  /// frame collects independent data.
+  static void reset() {
+    _buffer = _BlockBuffer();
+  }
+}
+
+/// Provides [start], [end], and [duration] of a named block of code, timed by
+/// [FlutterTimeline].
+@immutable
+final class TimedBlock {
+  /// Creates a timed block of code from a [name], [start], and [end].
+  ///
+  /// The [name] should be sufficiently unique and descriptive for someone to
+  /// easily tell which part of code was measured.
+  const TimedBlock({
+    required this.name,
+    required this.start,
+    required this.end,
+  }) : assert(end >= start, 'The start timestamp must not be greater than the end timestamp.');
+
+  /// A readable label for a block of code that was measured.
+  ///
+  /// This field should be sufficiently unique and descriptive for someone to
+  /// easily tell which part of code was measured.
+  final String name;
+
+  /// The timestamp in microseconds that marks the beginning of the measure block
+  /// of code.
+  final double start;
+
+  /// The timestamp in microseconds that marks the end of the measure block of
+  /// code.
+  final double end;
+
+  /// How long the measured block of code took to execute in microseconds.
+  double get duration => end - start;
+
+  @override
+  String toString() {
+    return 'TimedBlock($name, $start, $end, $duration)';
+  }
+}
+
+/// Provides aggregated results for timings collected by [FlutterTimeline].
+@immutable
+final class AggregatedTimings {
+  /// Creates aggregated timings for the provided timed blocks.
+  AggregatedTimings(this.timedBlocks) {
+    final Map<String, (double, int)> aggregate = <String, (double, int)>{};
+    for (final TimedBlock block in timedBlocks) {
+      final (double, int) previousValue = aggregate.putIfAbsent(block.name, () => (0, 0));
+      aggregate[block.name] = (previousValue.$1 + block.duration, previousValue.$2 + 1);
+    }
+
+    aggregatedBlocks = aggregate.entries.map<AggregatedTimedBlock>(
```

Comment:
Should these maybe only be computed if someone accesses aggregatedBlocks and/or getAggregated?

Reply:
Done.